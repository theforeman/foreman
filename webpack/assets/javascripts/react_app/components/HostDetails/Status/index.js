import PropTypes from 'prop-types';
import React, { useCallback } from 'react';
import { Alert } from '@patternfly/react-core';
import Skeleton from 'react-loading-skeleton';

const Status = ({ status }) => {
  const statusToLabel = useCallback(() => {
    switch (status) {
      case 'OK':
        return ['success', 'Host Status - OK'];
      case 'Error':
        return ['danger', 'Host Status - ERROR'];
      case 'Warning':
        return ['warning', 'Host Status - WARNING '];
      default:
        return ['info', null];
    }
  }, [status]);
  const [variant, title] = statusToLabel();
  return <Alert variant={variant} title={title || <Skeleton />} />;
};

Status.propTypes = {
  status: PropTypes.string.isRequired,
};

export default Status;
