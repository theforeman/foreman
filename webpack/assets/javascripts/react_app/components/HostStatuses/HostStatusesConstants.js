export const GLOBAL_STATUS_OK = 0;
export const GLOBAL_STATUS_WARN = 1;
export const GLOBAL_STATUS_ERROR = 2;
export const HOST_STATUSES_KEY = 'HOST_STATUSES';
export const API_OPTIONS = { key: HOST_STATUSES_KEY };
