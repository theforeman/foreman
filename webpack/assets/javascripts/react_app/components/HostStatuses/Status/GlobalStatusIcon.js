import React from 'react';
import PropTypes from 'prop-types';
import {
  CheckCircleIcon,
  ExclamationTriangleIcon,
  ExclamationCircleIcon,
  QuestionCircleIcon,
} from '@patternfly/react-icons';
import {
  GLOBAL_STATUS_OK,
  GLOBAL_STATUS_WARN,
  GLOBAL_STATUS_ERROR,
} from '../HostStatusesConstants';

const GlobalStatusIcon = ({ status, ...props }) => {
  switch (status) {
    case GLOBAL_STATUS_OK:
      return (
        <CheckCircleIcon
          style={{ fill: 'var(--pf-global--success-color--100)' }}
          {...props}
        />
      );
    case GLOBAL_STATUS_WARN:
      return (
        <ExclamationTriangleIcon
          style={{ fill: 'var(--pf-global--warning-color--100)' }}
          {...props}
        />
      );
    case GLOBAL_STATUS_ERROR:
      return (
        <ExclamationCircleIcon
          style={{ fill: 'var(--pf-global--danger-color--100)' }}
          {...props}
        />
      );
    default:
      return (
        <QuestionCircleIcon
          style={{ fill: 'var(--pf-global--info-color--200)' }}
          {...props}
        />
      );
  }
};

GlobalStatusIcon.propTypes = {
  status: PropTypes.number,
};

GlobalStatusIcon.defaultProps = {
  status: undefined,
};

export default GlobalStatusIcon;
