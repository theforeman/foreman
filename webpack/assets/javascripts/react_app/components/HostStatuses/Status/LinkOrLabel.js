import React from 'react';
import PropTypes from 'prop-types';

const LinkOrLabel = ({ label, path }) => {
  if (path) {
    return <a href={path}>{label}</a>;
  }
  return <span>{label}</span>;
};

LinkOrLabel.propTypes = {
  label: PropTypes.string.isRequired,
  path: PropTypes.string,
};

LinkOrLabel.defaultProps = {
  path: undefined,
};

export default LinkOrLabel;
