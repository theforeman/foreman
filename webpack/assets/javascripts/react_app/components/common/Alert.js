import React from 'react';
import Icon from './Icon';
import { ALERT_CSS } from '../../constants';

const Alert = (props) => {
  let classNames = ALERT_CSS[props.type];

  if (props.css) {
    classNames += ' ' + props.css;
  }

  if (props.dismissable) {
    classNames += ' ' + ALERT_CSS.dismissable;
  }

  const closeBtn = (
    <button className="close" data-dismiss="alert" aria-hidden="true">
      <Icon type="close" />
    </button>
  );

  return (
    <div className={classNames}>
      {props.dismissable && closeBtn}
      {props.children}
    </div>
  );
};

Alert.defaultProps = { dismissable: false, css: null, type: 'info' };
export default Alert;
