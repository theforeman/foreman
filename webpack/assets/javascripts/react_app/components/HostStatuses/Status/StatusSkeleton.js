import React from 'react';
import Skeleton from 'react-loading-skeleton';
import { Card, CardHeader, Grid, GridItem } from '@patternfly/react-core';
import GlobalStatusIcon from './GlobalStatusIcon';

const StatusSkeleton = () => (
  <GridItem sm={12} xl2={6}>
    <Card className="pf-c-alert pf-m-success">
      <CardHeader
        onExpand={() => {}}
        toggleButtonProps={{
          id: 'toggle-button',
          'aria-label': 'Details',
          'aria-labelledby': 'titleId toggle-button',
          'aria-expanded': false,
        }}
      >
        <Grid className="w-100" hasGutter>
          <GridItem span={1} rowSpan={2} className="text-center">
            <GlobalStatusIcon status={0} />
          </GridItem>
          <GridItem span={5} style={{ fontSize: '1.5em' }}>
            <Skeleton />
          </GridItem>
          <GridItem span={2} rowSpan={2} className="text-center">
            <div style={{ fontSize: '1.5em' }}>
              <GlobalStatusIcon status={0} />
            </div>
            <Skeleton />
            <br />
            <Skeleton />
          </GridItem>
          <GridItem span={2} rowSpan={2} className="text-center">
            <div style={{ fontSize: '1.5em' }}>
              <GlobalStatusIcon status={1} />
            </div>
            <Skeleton />
            <br />
            <Skeleton />
          </GridItem>
          <GridItem span={2} rowSpan={2} className="text-center">
            <div style={{ fontSize: '1.5em' }}>
              <GlobalStatusIcon status={2} />
            </div>
            <Skeleton />
            <br />
            <Skeleton />
          </GridItem>
          <GridItem span={5}>
            <Skeleton />
          </GridItem>
        </Grid>
      </CardHeader>
    </Card>
  </GridItem>
);

export default StatusSkeleton;
