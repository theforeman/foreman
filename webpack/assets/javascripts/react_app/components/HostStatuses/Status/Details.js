import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Table, TableHeader, TableBody } from '@patternfly/react-table';
import GlobalStatusIcon from './GlobalStatusIcon';
import LinkOrLabel from './LinkOrLabel';
import { translate as __ } from '../../../common/I18n';

const Details = ({ data }) => {
  const columns = ['', __('Total'), __('Owned')];
  const rows = data.map(
    ({
      label,
      total,
      owned,
      global_status: globalStatus,
      total_path: totalPath,
      owned_path: ownedPath,
    }) => [
      {
        title: (
          <Fragment>
            <GlobalStatusIcon status={globalStatus} /> {label}
          </Fragment>
        ),
      },
      { title: <LinkOrLabel path={totalPath} label={total.toString()} /> },
      { title: <LinkOrLabel path={ownedPath} label={owned.toString()} /> },
    ]
  );

  return (
    <Table
      aria-label="Host Statuses"
      variant="compact"
      cells={columns}
      rows={rows}
    >
      <TableHeader />
      <TableBody />
    </Table>
  );
};

Details.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      total: PropTypes.number.isRequired,
      owned: PropTypes.number.isRequired,
      global_status: PropTypes.number,
      total_path: PropTypes.string,
      owned_path: PropTypes.string,
    })
  ),
};

Details.propTypes = {
  data: PropTypes.array,
};

Details.defaultProps = {
  data: [],
};

export default Details;
