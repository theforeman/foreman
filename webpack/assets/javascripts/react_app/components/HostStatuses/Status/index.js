import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import { asMutable } from 'seamless-immutable';
import classNames from 'classnames';
import {
  Card,
  CardHeader,
  CardExpandableContent,
  Grid,
  GridItem,
} from '@patternfly/react-core';
import GlobalStatusIcon from './GlobalStatusIcon';
import { sprintf, translate as __ } from '../../../common/I18n';
import Details from './Details';
import LinkOrLabel from './LinkOrLabel';
import {
  selectGlobalStatus,
  selectHostStatusDetails,
  selectHostStatusDescription,
  selectHostStatusCounter,
  selectHostStatusTotalPaths,
  selectHostStatusOwnedPaths,
} from '../HostStatusesSelectors';

import './Status.scss';

const Status = ({ name }) => {
  const [expanded, setExpanded] = useState(false);

  const globalStatus = useSelector(state => selectGlobalStatus(state, name));
  const details = useSelector(state => selectHostStatusDetails(state, name));
  const description = useSelector(state =>
    selectHostStatusDescription(state, name)
  );
  const { okTotalPath, warnTotalPath, errorTotalPath } = useSelector(state =>
    selectHostStatusTotalPaths(state, name)
  );
  const { okOwnedPath, warnOwnedPath, errorOwnedPath } = useSelector(state =>
    selectHostStatusOwnedPaths(state, name)
  );

  const {
    unknown: { total: unknownTotalCount, owned: unknownOwnedCount },
    ok: { total: okTotalCount, owned: okOwnedCount },
    warn: { total: warnTotalCount, owned: warnOwnedCount },
    error: { total: errorTotalCount, owned: errorOwnedCount },
  } = useSelector(state => selectHostStatusCounter(state, name));

  const cardClass = classNames('pf-c-alert', {
    'pf-m-danger': errorTotalCount,
    'pf-m-warning': !errorTotalCount && warnTotalCount,
    'pf-m-success': !errorTotalCount && !warnTotalCount,
  });

  const unknownStatusesPresent = !!unknownTotalCount || !!unknownOwnedCount;

  return (
    <GridItem sm={12} xl2={6}>
      <Card className={cardClass} isExpanded={expanded} isHoverable>
        <CardHeader
          onExpand={(_event, _id) => setExpanded(!expanded)}
          toggleButtonProps={{
            id: 'toggle-button',
            'aria-label': 'Details',
            'aria-labelledby': 'titleId toggle-button',
            'aria-expanded': expanded,
          }}
        >
          <Grid className="w-100" hasGutter>
            <GridItem span={1} rowSpan={2} style={{ fontSize: '2.2em' }}>
              <GlobalStatusIcon status={globalStatus} />
            </GridItem>
            <GridItem
              span={unknownStatusesPresent ? 3 : 5}
              style={{ fontSize: '1.5em' }}
            >
              {name}
            </GridItem>
            {unknownStatusesPresent && (
              <GridItem
                span={2}
                rowSpan={2}
                className="status-count text-center"
              >
                <div style={{ fontSize: '1.5em' }}>
                  <GlobalStatusIcon />
                </div>
                <LinkOrLabel
                  label={sprintf(__('Total: %s'), unknownTotalCount)}
                />
                <LinkOrLabel
                  label={sprintf(__('Owned: %s'), unknownOwnedCount)}
                />
              </GridItem>
            )}
            <GridItem span={2} rowSpan={2} className="status-count text-center">
              <div style={{ fontSize: '1.5em' }}>
                <GlobalStatusIcon status={0} />
              </div>
              <LinkOrLabel
                path={okTotalPath}
                label={sprintf(__('Total: %s'), okTotalCount)}
              />
              <LinkOrLabel
                path={okOwnedPath}
                label={sprintf(__('Owned: %s'), okOwnedCount)}
              />
            </GridItem>
            <GridItem span={2} rowSpan={2} className="status-count text-center">
              <div style={{ fontSize: '1.5em' }}>
                <GlobalStatusIcon status={1} />
              </div>
              <LinkOrLabel
                path={warnTotalPath}
                label={sprintf(__('Total: %s'), warnTotalCount)}
              />
              <LinkOrLabel
                path={warnOwnedPath}
                label={sprintf(__('Owned: %s'), warnOwnedCount)}
              />
            </GridItem>
            <GridItem span={2} rowSpan={2} className="status-count text-center">
              <div style={{ fontSize: '1.5em' }}>
                <GlobalStatusIcon status={2} />
              </div>
              <LinkOrLabel
                path={errorTotalPath}
                label={sprintf(__('Total: %s'), errorTotalCount)}
              />
              <LinkOrLabel
                path={errorOwnedPath}
                label={sprintf(__('Owned: %s'), errorOwnedCount)}
              />
            </GridItem>
            <GridItem span={5}>{description}</GridItem>
          </Grid>
        </CardHeader>
        <CardExpandableContent>
          {details.length > 0 ? (
            <Details data={asMutable(details)} />
          ) : (
            __('Nothing to show')
          )}
        </CardExpandableContent>
      </Card>
    </GridItem>
  );
};

Status.propTypes = {
  name: PropTypes.string.isRequired,
};

export default Status;
