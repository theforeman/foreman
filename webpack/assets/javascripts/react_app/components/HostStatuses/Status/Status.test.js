import Status from '../Status';
import { HOST_STATUSES_KEY } from '../HostStatusesConstants';
import { store } from '../HostStatuses.fixtures.js'
import { testComponentSnapshotsWithFixtures } from '../../../common/testHelpers';

jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockImplementation(selector => selector()),
}));

jest.mock('../HostStatusesSelectors.js', () => {
  const { get } = require('lodash');
  const { HOST_STATUSES_KEY } = require('../HostStatusesConstants')
  const { store } = require('../HostStatuses.fixtures.js');

  const status = get(store, ['API', HOST_STATUSES_KEY, 'response', 'results', '0'])
  const details = get(status, 'details')

  return {
    selectGlobalStatus: jest.fn().mockReturnValue(2),
    selectHostStatusDetails: jest.fn().mockReturnValue(details),
    selectHostStatusDescription: jest.fn().mockReturnValue(get(status, 'description')),
    selectHostStatusTotalPaths: jest.fn().mockReturnValue({
      okTotalPath: get(status, 'ok_total_path'),
      warnTotalPath: get(status, 'warn_total_path'),
      errorTotalPath: get(status, 'error_total_path'),
    }),
    selectHostStatusOwnedPaths: jest.fn().mockReturnValue({
      errorOwnedPath: get(status, 'ok_owned_path'),
      okOwnedPath: get(status, 'warn_owned_path'),
      warnOwnedPath: get(status, 'error_owned_path'),
    }),
    selectHostStatusCounter: jest.fn().mockReturnValue({
      ok: {
        "owned": details.filter(({ global_status: gs }) => gs == 0).map(({ owned }) => owned).reduce((a, b) => a + b),
        "total": details.filter(({ global_status: gs }) => gs == 0).map(({ total }) => total).reduce((a, b) => a + b),
      },
      warn: {
        "owned": details.filter(({ global_status: gs }) => gs == 1).map(({ owned }) => owned).reduce((a, b) => a + b),
        "total": details.filter(({ global_status: gs }) => gs == 1).map(({ total }) => total).reduce((a, b) => a + b),
      },
      error: {
        "owned": details.filter(({ global_status: gs }) => gs == 2).map(({ owned }) => owned).reduce((a, b) => a + b),
        "total": details.filter(({ global_status: gs }) => gs == 2).map(({ total }) => total).reduce((a, b) => a + b),
      },
      unknown: {
        "owned": details.filter(({ global_status: gs }) => gs == null).map(({ owned }) => owned).reduce((a, b) => (a + b), 0),
        "total": details.filter(({ global_status: gs }) => gs == null).map(({ total }) => total).reduce((a, b) => (a + b), 0),
      },
    }),
  }
});

const fixtures = {
  'renders Status': { name: store.API[HOST_STATUSES_KEY].response.results[0].name },
};
describe('Status', () => {
  testComponentSnapshotsWithFixtures(Status, fixtures);
});
