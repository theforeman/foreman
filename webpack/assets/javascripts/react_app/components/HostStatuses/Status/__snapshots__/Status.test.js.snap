// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status renders Status 1`] = `
<GridItem
  sm={12}
  xl2={6}
>
  <Card
    className="pf-c-alert pf-m-danger"
    isExpanded={false}
    isHoverable={true}
  >
    <CardHeader
      onExpand={[Function]}
      toggleButtonProps={
        Object {
          "aria-expanded": false,
          "aria-label": "Details",
          "aria-labelledby": "titleId toggle-button",
          "id": "toggle-button",
        }
      }
    >
      <Grid
        className="w-100"
        hasGutter={true}
      >
        <GridItem
          rowSpan={2}
          span={1}
          style={
            Object {
              "fontSize": "2.2em",
            }
          }
        >
          <GlobalStatusIcon
            status={2}
          />
        </GridItem>
        <GridItem
          span={5}
          style={
            Object {
              "fontSize": "1.5em",
            }
          }
        >
          Status Name
        </GridItem>
        <GridItem
          className="status-count text-center"
          rowSpan={2}
          span={2}
        >
          <div
            style={
              Object {
                "fontSize": "1.5em",
              }
            }
          >
            <GlobalStatusIcon
              status={0}
            />
          </div>
          <LinkOrLabel
            label="Total: 3"
            path="/hosts?search=status+%3D+ok"
          />
          <LinkOrLabel
            label="Owned: 1"
            path="/hosts?search=owner+%3D+current_user+AND+%28status+%3D+warn"
          />
        </GridItem>
        <GridItem
          className="status-count text-center"
          rowSpan={2}
          span={2}
        >
          <div
            style={
              Object {
                "fontSize": "1.5em",
              }
            }
          >
            <GlobalStatusIcon
              status={1}
            />
          </div>
          <LinkOrLabel
            label="Total: 7"
            path="/hosts?search=status+%3D+warn"
          />
          <LinkOrLabel
            label="Owned: 2"
            path="/hosts?search=owner+%3D+current_user+AND+%28status+%3D+error"
          />
        </GridItem>
        <GridItem
          className="status-count text-center"
          rowSpan={2}
          span={2}
        >
          <div
            style={
              Object {
                "fontSize": "1.5em",
              }
            }
          >
            <GlobalStatusIcon
              status={2}
            />
          </div>
          <LinkOrLabel
            label="Total: 5"
            path="/hosts?search=status+%3D+error"
          />
          <LinkOrLabel
            label="Owned: 0"
            path="/hosts?search=owner+%3D+current_user+AND+%28status+%3D+ok"
          />
        </GridItem>
        <GridItem
          span={5}
        >
          Description of the status
        </GridItem>
      </Grid>
    </CardHeader>
    <CardExpandableContent>
      <Details
        data={
          Array [
            Object {
              "global_status": 0,
              "label": "OK",
              "owned": 1,
              "owned_path": "/hosts?search=owner+%3D+current_user+AND+%28status+%3D+ok",
              "total": 3,
              "total_path": "/hosts?search=status+%3D+ok",
            },
            Object {
              "global_status": 1,
              "label": "Warning",
              "owned": 2,
              "owned_path": "/hosts?search=owner+%3D+current_user+AND+%28status+%3D+warn",
              "total": 7,
              "total_path": "/hosts?search=status+%3D+warn",
            },
            Object {
              "global_status": 2,
              "label": "Error",
              "owned": 0,
              "owned_path": "/hosts?search=owner+%3D+current_user+AND+%28status+%3D+error",
              "total": 5,
              "total_path": "/hosts?search=status+%3D+error",
            },
          ]
        }
      />
    </CardExpandableContent>
  </Card>
</GridItem>
`;
